snow/validators/manager.go

```diff
@@ -49,27 +108,177 @@ type manager struct {
 
 	// Key: Subnet ID
 	// Value: The validators that validate the subnet
-	subnetToVdrs map[ids.ID]Set
+	subnetToVdrs map[ids.ID]*vdrSet
+}
+
+func (m *manager) AddStaker(subnetID ids.ID, nodeID ids.NodeID, pk *bls.PublicKey, txID ids.ID, weight uint64) error {
+	// Use write lock in case we need to create a new set
+	m.lock.Lock()
+	set, exists := m.subnetToVdrs[subnetID]
+	if !exists {
+		set = newSet()
+		m.subnetToVdrs[subnetID] = set
+	}
+	m.lock.Unlock()
+
+	// Use read lock for all other operations
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+
+	return set.Add(nodeID, pk, txID, weight)
```

Comment:
![image](https://github.com/ava-labs/avalanchego/assets/22109487/9db3b374-2d18-40a1-ae1c-48d5a60ea244)

I think there is some racy behavior here.

If we have 2 goroutines both executing: `AddStaker` followed by `RemoveWeight` on the same subnet concurrently - then I think it's possible for the first call to `RemoveWeight` to overlap poorly with the second call to `AddStaker` in such a way that the addition is dropped.

Reply:
yea correct. changed with full locks.